### PR TITLE
distsql: add unit test for distsql.go

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

**Comment by shenli** [comment]
How about the test coverage now?

**Comment by zz-jason** [comment]
@shenli current coverage is 32.8%, still too low, I'm working on it.

File: distsql/distsql_test.go [hunk 1]
```diff
@@ -15,31 +15,115 @@ package distsql
...
+	c.Assert(chk.NumRows(), Equals, 0)
 }
 
-func (resp *mockResponse) Next(ctx context.Context) ([]byte, error) {
+type mockStore struct{ client *mockClient }
+type mockClient struct{}
+type mockResponse struct{ count int }
+type mockResultSubset struct{ data []byte }
+
+// functions for mockStore.
```
Comment (zimulala) [inline distsql/distsql_test.go]: Could we put this logic into util/mock like `context`?
I see the same logic in logical_plan_test.go.

File: distsql/distsql_test.go [hunk 2]
```diff
@@ -15,31 +15,115 @@ package distsql
...
 	"github.com/pingcap/tipb/go-tipb"
 	"golang.org/x/net/context"
 )
 
-type mockResponse struct {
-	count int
+func (s *testSuite) TestSelect(c *C) {
+	keyRanges := []kv.KeyRange{
+		{
+			StartKey: kv.Key{0x74, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf, 0x5f, 0x72, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0},
```
Comment (zimulala) [inline distsql/distsql_test.go]: Do we need to comment out this key? It's hard to understand.

**Comment by zimulala** [comment]
Please fix CI.

**Comment by zz-jason** [comment]
test coverage

| file name | new coverage | old coverage |
|----------|---------------|--------------|
| distsql.go | 83.3% | 0% |
| request_builder.go | 90.0% | 90.0% |
| stream.go | 71.1% | 0% |

File: distsql/distsql_test.go (hunk 2 above)
Comment (zz-jason) [reply distsql/distsql_test.go]: OK

**Comment by zz-jason** [comment]
@zimulala @tiancaiamao PTAL

File: distsql/request_builder_test.go [hunk 3]
```diff
@@ -42,6 +45,23 @@ func TestT(t *testing.T) {
...
+	ctx := mock.NewContext()
+	ctx.Store = &mock.Store{&mock.Client{&mockResponse{}}}
+	s.sctx = ctx
+}
+
+func (s *testSuite) TearDownSuite(c *C) {
+}
+
+func (s *testSuite) SetUpTest(c *C) {
+	testleak.BeforeTest()
```
Comment (tiancaiamao) [inline distsql/request_builder_test.go]: Where is `testleak.AfterTest()()` called?

Comment (zz-jason) [reply distsql/request_builder_test.go]: added in `TearDownTest`

**Comment by tiancaiamao** [comment]
LGTM

**Comment by zz-jason** [comment]
@zimulala PTAL

File: distsql/distsql_test.go [hunk 4]
```diff
@@ -14,32 +14,239 @@
...
+		numAllRows += chk.NumRows()
+		if chk.NumRows() == 0 {
+			break
+		}
+	}
+	c.Assert(numAllRows, Equals, 2)
+	err = response.Close()
+	c.Assert(err, IsNil)
+
+	// Test Next.
```
Comment (zhexuany) [inline distsql/distsql_test.go]: since Next is being removing, we're better only cover NextChunk()

File: distsql/distsql_test.go [hunk 5]
```diff
@@ -14,32 +14,239 @@
...
+	c.Assert(len(bytes), Equals, 14)
+
+	err = response.Close()
+	c.Assert(err, IsNil)
 }
 
+type mockResponse struct{ count int }
+type mockResultSubset struct{ data []byte }
+
+// functions for mockResponse.
```
Comment (zimulala) [inline distsql/distsql_test.go]: The format of the comment is as follows:
`Close ...`

File: distsql/distsql_test.go [hunk 6]
```diff
@@ -14,32 +14,239 @@
...
+	}
+	respBytes, err := respPB.Marshal()
 	if err != nil {
 		panic(err)
 	}
-	return b
+	return &mockResultSubset{respBytes}, nil
 }
+
+// functions for mockResultSubset.
```
Comment (zimulala) [inline distsql/distsql_test.go]: Ditto.

**Comment by zz-jason** [comment]
@zimulala @zhexuany PTAL

File: distsql/distsql_test.go [hunk 7]
```diff
@@ -14,32 +14,182 @@
...
+	resp.count = 0
 	return nil
 }
 
-func mockSubresult() []byte {
-	resp := new(tipb.SelectResponse)
-	b, err := resp.Marshal()
+// Next implements kv.Response interface.
+func (resp *mockResponse) Next(ctx context.Context) (kv.ResultSubset, error) {
+	if resp.count == 2 {
```
Comment (zhexuany) [inline distsql/distsql_test.go]: This snippet confuses me. Why are u directly returning nil value to its caller when `count` is 2? may be you need comment it?

Comment (zz-jason) [reply distsql/distsql_test.go]: It's mocked, we only return 2 rows for convenience. However you can return as many rows as you wish.

**Pull request merged**